File: 2021/12/12.hs
import System.IO (readFile)

data Line = Kategori { level :: Int } | Gave { level :: Int } deriving (Show)

isKategori :: Line -> Bool
isKategori (Kategori _) = True
isKategori _ = False

parseLine :: String -> Line
parseLine line = let (dashes, (t:_)) = break (/= '-') line
                     level = length dashes
                 in  case t of 'K' -> Kategori level
                               'G' -> Gave level

removeEmptyCategories :: [Line] -> [Line]
removeEmptyCategories lines = rec lines []
    where rec [] acc = reverse acc
          rec (g@(Gave _):xs) acc = rec xs (g:acc)
          rec (a@(Kategori l) : b : xs) (a':acc)
              | level b <= l = rec (a':b:xs) acc
              | otherwise    = rec (b:xs) (a:a':acc)

main = do
    rawLines <- lines <$> readFile "task.txt"
    let lines = map parseLine rawLines

    print
        . length
        . filter isKategori
        . removeEmptyCategories
        $ lines

File: 2021/12/12.js
"use strict";

const fs = require("fs");

fs.promises.readFile("./task.txt", "utf-8")
	.then(content => content.split("\n").slice(0,-1))
	.then(lines => lines.map(line => {
		const [, dashes, type, name] = /^(-*)(K|G) (.*)$/.exec(line);
		return {
			level: dashes.length,
			type,
			name
		};
	}))
	.then(lines => {
		let i = 0;
		while (i < lines.length-1) {
			if (lines[i].type === "K" && lines[i+1].level <= lines[i].level) {
				lines.splice(i, 1);
				i--;
			} else {
				i++;
			}
		}

		return lines;
	})
	.then(lines => lines.filter(({ type }) => type === "K"))
	.then(lines => console.log(lines.length));
